File: SH/graph/787. Cheapest_Flights_within_k_steps.c
#include <stdlib.h>
#include <limits.h>

struct Node {
    int dest;
    int price;
    struct Node* next;
};

struct Graph {
    int V;
    struct Node** adjList;
};

struct Node* createNode(int dest, int price) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->dest = dest;
    newNode->price = price;
    newNode->next = NULL;
    return newNode;
}

struct Graph* createGraph(int V) {
    struct Graph* graph = (struct Graph*)malloc(sizeof(struct Graph));
    graph->V = V;
    graph->adjList = (struct Node**)malloc(V * sizeof(struct Node*));
    for (int i = 0; i < V; ++i)
        graph->adjList[i] = NULL;
    return graph;
}

void addEdge(struct Graph* graph, int src, int dest, int price) {
    struct Node* newNode = createNode(dest, price);
    newNode->next = graph->adjList[src];
    graph->adjList[src] = newNode;
}

int findCheapestPrice(int n, int** flights, int flightsSize, int* flightsColSize, int src, int dst, int k) {
    struct Graph* graph = createGraph(n);
    int* visited = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; ++i)
        visited[i] = INT_MAX;
    visited[src] = 0;
    
    for (int i = 0; i < flightsSize; ++i)
        addEdge(graph, flights[i][0], flights[i][1], flights[i][2]);
    
    struct Node* queue = createNode(src, 0);
    k++;
    
    while (queue && k--) {
        int count = 0;
        struct Node* tempQueue = createNode(-1, -1);
        while (queue) {
            struct Node* curr = queue;
            queue = queue->next;
            int currNode = curr->dest;
            int currPrice = curr->price;
            struct Node* adjNode = graph->adjList[currNode];
            while (adjNode) {
                int newPrice = currPrice + adjNode->price;
                if (newPrice < visited[adjNode->dest]) {
                    visited[adjNode->dest] = newPrice;
                    struct Node* newNode = createNode(adjNode->dest, newPrice);
                    newNode->next = tempQueue->next;
                    tempQueue->next = newNode;
                }
                adjNode = adjNode->next;
            }
            free(curr);
            count++;
        }
        free(queue);
        queue = tempQueue->next;
    }
    
    return visited[dst] == INT_MAX ? -1 : visited[dst];
}

File: SH/graph/787. Cheapest_Flights_within_k_steps.py
'''
787. Cheapest Flights Within K Stops
Medium
Topics
Companies
There are n cities connected by some number of flights. You are given an array flights where flights[i] = [fromi, toi, pricei] indicates that there is a flight from city fromi to city toi with cost pricei.

You are also given three integers src, dst, and k, return the cheapest price from src to dst with at most k stops. If there is no such route, return -1.

 
Example 1:
Input: n = 4, flights = [[0,1,100],[1,2,100],[2,0,100],[1,3,600],[2,3,200]], src = 0, dst = 3, k = 1
Output: 700
Explanation:
The graph is shown above.
The optimal path with at most 1 stop from city 0 to 3 is marked in red and has cost 100 + 600 = 700.
Note that the path through cities [0,1,2,3] is cheaper but is invalid because it uses 2 stops.
Example 2:
Input: n = 3, flights = [[0,1,100],[1,2,100],[0,2,500]], src = 0, dst = 2, k = 1
Output: 200
Explanation:
The graph is shown above.
The optimal path with at most 1 stop from city 0 to 2 is marked in red and has cost 100 + 100 = 200.
Example 3:
Input: n = 3, flights = [[0,1,100],[1,2,100],[0,2,500]], src = 0, dst = 2, k = 0
Output: 500
Explanation:
The graph is shown above.
The optimal path with no stops from city 0 to 2 is marked in red and has cost 500.
'''
from typing import List
from collections import deque, defaultdict
import heapq # as this needs to prioritize the step, better not to use heapq

inf = 1e10

class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        # adj-list
        edges = defaultdict(list)
        for s, d, price in flights:
            edges[s].append((d, price))

        q = deque([(src, 0)]) # [node, price]: it takes 0 price to reach src
        visited = [inf] * n
        visited[src] = 0
        k += 1

        while k>0 and q:
            size = len(q)
            while size > 0:
                curr, price = q.popleft()
                # for all neighbors
                for node, price_ in edges[curr]:
                    # print(node, price)
                    new_price = price_ + price
                    if new_price < visited[node]:
                        visited[node] = new_price
                        q.append((node, new_price))
                size -= 1
            k -= 1

        return visited[dst] if visited[dst] != inf else -1



if __name__ == "__main__":
    print("743. Network_delay_time.py")

    solution = Solution()
    print("\nCase 1")
    flights = [[0,1,100],[1,2,100],[2,0,100],[1,3,600],[2,3,200]]
    src = 0
    dst = 3
    n = 4
    k = 1
    res = solution.findCheapestPrice(n, flights, src, dst, k)  
    ans = 700
    print(res, "\n  ==> check: ", res==ans)

    print("\nCase 2")
    flights = [[0,1,100],[1,2,100],[0,2,500]] 
    src = 0
    dst = 2
    n = 3
    k = 1
    res = solution.findCheapestPrice(n, flights, src, dst, k) 
    ans = 200
    print(res, "\n  ==> check: ", res==ans)

    print("\nCase 3")
    flights = [[0,1,100],[1,2,100],[0,2,500]]
    src = 0
    dst = 2
    n = 3
    k = 0
    res = solution.findCheapestPrice(n, flights, src, dst, k) 
    ans = 500
    print(res, "\n  ==> check: ", res==ans)
